File: 백준/Gold/11567. 선진이의 겨울 왕국/선진이의 겨울 왕국.py
# https://www.acmicpc.net/problem/11567
import sys
from collections import deque
input = sys.stdin.readline
go = ((-1, 0), (1, 0), (0, 1), (0, -1))


def bfs() -> bool:
    q = deque()
    q.append((r1, c1))

    while q:
        x, y = q.popleft()

        for a, b in go:
            nx, ny = x+a, y+b

            if nx < 0 or nx >= n or ny < 0 or ny >= m:
                continue
            if board[nx][ny]:
                if (nx, ny) == (r2, c2):
                    return True
                else:
                    continue

            board[nx][ny] = 1
            q.append((nx, ny))


n, m = map(int, input().split())
board = [[1 if x == 'X' else 0 for x in input().rstrip()] for _ in range(n)]
r1, c1 = map(lambda x: int(x)-1, input().split())
r2, c2 = map(lambda x: int(x)-1, input().split())

print('YES' if bfs() else 'NO')

File: 백준/Gold/11567. 선진이의 겨울 왕국/선진이의 겨울 왕국.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Queue;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int[] nm = Arrays.stream(br.readLine().split(" "))
                .mapToInt(Integer::parseInt)
                .toArray();

        int n = nm[0];
        int m = nm[1];

        boolean[][] board = new boolean[n][m];
        for (int i = 0; i < n; i++) {
            String line = br.readLine();
            for (int j = 0; j < m; j++) {
                board[i][j] = line.charAt(j) == 'X';
            }
        }

        int[] start = Arrays.stream(br.readLine().split(" "))
                .mapToInt(Integer::parseInt)
                .toArray();
        int[] destination = Arrays.stream(br.readLine().split(" "))
                .mapToInt(Integer::parseInt)
                .toArray();

        boolean ans = bfs(board, n, m, start, destination);
        System.out.println(ans ? "YES" : "NO");
    }

    private static boolean bfs(boolean[][] board, int n, int m, int[] start, int[] destination) {
        int[] goX = {-1, 1, 0, 0};
        int[] goY = {0, 0, -1, 1};

        int sx = start[0] - 1;
        int sy = start[1] - 1;
        int dx = destination[0] - 1;
        int dy = destination[1] - 1;

        Queue<int[]> q = new ArrayDeque<>();
        q.add(new int[]{sx, sy});

        while (!q.isEmpty()) {
            int[] first = q.poll();
            int x = first[0];
            int y = first[1];

            for (int d = 0; d < 4; d++) {
                int nx = x + goX[d];
                int ny = y + goY[d];

                if (nx < 0 || nx >= n || ny < 0 || ny >= m)
                    continue;

                if (board[nx][ny]) {
                    if (nx == dx && ny == dy)
                        return true;
                    else
                        continue;
                }

                board[nx][ny] = true;
                q.add(new int[]{nx, ny});
            }
        }

        return false;
    }
}